pyro/infer/autoguide/effect.py

```diff
@@ -0,0 +1,223 @@
+        return value
+
+
+class AutoNormalMessenger(AutoMessenger):
+    """
+    EXPERIMENTAL Automatic :class:`~pyro.infer.effect_elbo.GuideMessenger` ,
+    intended for use with :class:`~pyro.infer.effect_elbo.Effect_ELBO` or
+    similar.
+
+    The mean-field posterior at any site is a transformed normal distribution.
+
+    Derived classes may override particular sites and use this simply as a
+    default, e.g.::
+
+        def model(data):
+            a = pyro.sample("a", dist.Normal(0, 1))
+            b = pyro.sample("b", dist.Normal(0, 1))
+            c = pyro.sample("c", dist.Normal(a + b, 1))
+            pyro.sample("obs", dist.Normal(c, 1), obs=data)
+
+        class MyGuideMessenger(AutoNormalMessenger):
+            def get_posterior(self, name, prior, upstream_values):
+                if name == "c":
+                    # Use a custom distribution at site c.
+                    bias = pyro.param("c_bias", lambda: torch.zeros(()))
+                    weight = pyro.param("c_weight", lambda: torch.ones(()),
+                                        constraint=constraints.positive)
+                    scale = pyro.param("c_scale", lambda: torch.ones(()),
+                                       constraint=constraints.positive)
+                    a = upstream_values["a"]
+                    b = upstream_values["b"]
+                    loc = bias + weight * (a + b)
+                    return dist.Normal(loc, scale)
+                # Fall back to mean field.
+                return super().get_posterior(name, prior, upstream_values)
+
+    Note that above we manually computed ``loc = bias + weight * (a + b)``.
+    Alternatively we could reuse the model-side computation by setting ``loc =
+    bias + weight * prior.loc``::
+
+        class MyGuideMessenger_v2(AutoNormalMessenger):
+            def get_posterior(self, name, prior, upstream_values):
+                if name == "c":
+                    # Use a custom distribution at site c.
+                    bias = pyro.param("c_bias", lambda: torch.zeros(()))
+                    scale = pyro.param("c_scale", lambda: torch.ones(()),
+                                       constraint=constraints.positive)
+                    weight = pyro.param("c_weight", lambda: torch.ones(()),
+                                        constraint=constraints.positive)
+                    loc = bias + weight * prior.loc
+                    return dist.Normal(loc, scale)
+                # Fall back to mean field.
+                return super().get_posterior(name, prior, upstream_values)
+    """
+
+    @pyro_method
+    def get_posterior(
+        self,
+        name: str,
+        prior: TorchDistribution,
+        upstream_values: Dict[str, torch.Tensor],
+    ) -> Union[TorchDistribution, torch.Tensor]:
+        with helpful_support_errors({"name": name, "fn": prior}):
+            transform = biject_to(prior.support)
+        loc, scale = self._get_params(name, prior)
```

Comment:
Correct, the intention of this simple guide is to be mean field.

Do you want to try contributing an `AutoHierchicalNormalMessenger` guide as a follow-up to this PR?  I tried to do something similar with `AutoRegressiveMessenger` below by sampling from the prior and then shifting in unconstrained space.  I was unsure how to implement a general `AutoHierarchicalNormalMessenger` because not all `prior` distributions have a `.mean` method, and even then it is the mean in unconstrained space that we care about. E.g. how do we deal with `Gamma` or `LogNormal` or `Beta` or `Dirichlet`?

Reply:
I understand your point about the distributions that don't have the mean. What are those distributions by the way?

I am thinking about this solution:
```python
loc, scale, weight = self._get_params(name, prior)
loc = loc + transform.inv(prior.loc) * weight
```
Does it make sense for all distributions that have the mean?

Reply:
Yes, I will do`AutoHierchicalNormalMessenger` PR - should I wait until this PR is merged?

Reply:
> What [distributions do not have a `.mean` method] by the way?

- Heavy tailed distributions may not have a mean, e.g. `Cauchy` and `Stable` have infinite variance and no defined mean
- Non-euclidean distributions such as `VonMises3D` and `ProjectedNormal` have no defined mean.
- Some complex distributions have no computable mean, e.g. `TransformedDistribution(Normal(...), MyNormalizingFlow)`.

> Does `prior.loc` make sense for all distributions that have the mean?

First I would opt for `prior.mean` rather than `prior.loc`, since e.g. `LogNormal(...).loc` isn't a mean, rather it is the mean of the pre-transformed normal. Second note that the transform of the constrained mean is not the same as the unconstrained mean or unconstrained median, e.g. for LogNormal, `mean = exp(loc + scale**2 / 2)` whereas `median = exp(loc)`.

I think your `.mean` idea is good enough in most cases, and for cases where it fails, users can subclass and define their own custom `.get_posterior()` methods.